Sources/BSON/Document.swift

```diff
@@ -1,6 +1,33 @@
 import Foundation
 import NIO
 
+/// This shared allocator instance should be used for all underlying `ByteBuffer` creation.
+private let BSON_ALLOCATOR = ByteBufferAllocator()
+
+extension ByteBuffer {
+    @discardableResult
+    internal mutating func writeCString(_ string: String) -> Int {
+        let written = self.writeString(string + "\0")
+        return written
+    }
+
+    internal mutating func readCString() throws -> String? {
+        var string = ""
+        for i in 0..<0xFFFFFE {
+            if let b = self.readBytes(length: 1) {
+                if b[0] == 0 {
+                    return string
+                }
+                guard let character = String(bytes: b, encoding: .utf8) else {
+                    throw InternalError(message: "Cannot decode CString, byte: \(b) at position \(i) as utf8")
+                }
+                string += character
+            }
+        }
+        throw InternalError(message: "Failed to read CString, possibly missing null terminator?")
```

Comment:
so the only other way this could fail is if your buffer is too long to be valid BSON, right?

Reply:
Yep. Although I'll add a condition actually incase the readIndex is in a position where you can't read anything at all.